pax-manifest/src/deserializer/mod.rs

```diff
@@ -34,76 +35,64 @@ impl Deserializer {
-        match unit {
-            "%" => Ok(IntoableLiteral::Percent(from_pax(str).unwrap())),
-            _ => Err(Error::UnsupportedMethod),
-        }
-    } else if let Ok(_ast) = PaxParser::parse(Rule::literal_number, str) {
-        Ok(IntoableLiteral::Numeric(from_pax(str).unwrap()))
-    } else {
-        Err(Error::UnsupportedMethod) //Not an IntoableLiteral
-    };
-
-    CACHED_VALUES_INTO.with(|cache| {
-        cache.borrow_mut().insert(str.to_string(), ret.clone());
-    });
-
-    ret
+/// Given type information T, this coerces the value of the PaxAny into the expected
+/// type if able, or returns an error
+pub fn from_pax_try_coerce<T: ToFromPaxAny + CoercionRules + Clone + 'static>(
+    str: &str,
+) -> std::result::Result<PaxAny, String>
+where
+    T: DeserializeOwned,
+{
+    from_pax::<T>(str)
+        .map_err(|e| format!("failed to deserialize: {:?}", e))
+        .and_then(|v| v.try_coerce::<T>())
 }
 
-/// Main entry-point for deserializing a type from Pax.
-pub fn from_pax<T: Clone + 'static>(str: &str) -> Result<T>
+fn from_pax<T: ToFromPaxAny + CoercionRules + Clone + 'static>(str: &str) -> Result<PaxAny>
 where
     T: DeserializeOwned,
 {
     if let Some(cached) = CACHED_VALUES.with(|cache| {
         let cache = cache.borrow();
         let option_cached_dyn_any = cache.get(str);
         // down cast val to T
-        if let Some(cached_dyn_any) = &option_cached_dyn_any {
-            let option_t_value: Option<&T> = cached_dyn_any.downcast_ref::<T>();
-            if let Some(data) = option_t_value {
-                return Some(data.clone());
-            }
+        if let Some(data) = &option_cached_dyn_any {
+            return Some(data.try_clone::<T>().unwrap());
         }
         None
     }) {
-        return Ok(cached.clone());
+        return Ok(cached);
+    }
+
+    // COMMENT-TO-REVIEWER:
+    // What do do here?
+    // I feel like these should be movable into the deserializer,
+    // but for some reason it doesn't work when I do so.
+    let type_id = TypeId::of::<T>();
+    if type_id != TypeId::of::<Color>()
+        && type_id != TypeId::of::<Percent>()
+        && type_id != TypeId::of::<Numeric>()
+    {
+        let ret = if let Ok(_ast) = PaxParser::parse(Rule::literal_color, str) {
+            Ok(from_pax::<Color>(str).unwrap())
+        } else if let Ok(ast) = PaxParser::parse(Rule::literal_number_with_unit, str) {
+            let _number = ast.clone().next().unwrap().as_str();
+            let unit = ast.clone().next().unwrap().as_str();
+            match unit {
+                "%" => Ok(from_pax::<Percent>(&str).unwrap()),
+                _ => Err(Error::UnsupportedMethod),
+            }
+        } else if let Ok(_ast) = PaxParser::parse(Rule::literal_number, str) {
+            Ok(from_pax::<Numeric>(str).unwrap())
+        } else {
+            Err(Error::UnsupportedMethod)
+        };
+        if let Ok(val) = ret {
+            return Ok(val);
+        }
```

Comment:
@warfaj This is the part I'm unsure about! Is there some way of integrating this into the deserializer in a better way?

Reply:
For context: These pre-checks were taken from the body of the IntoableLiteral parser.

Reply:
We can leave this for now and revisit when we turn PaxAny -> PaxValue.